pubsub/api/PubsubTest/PubsubTest.cs

```diff
@@ -42,15 +42,60 @@ public class PubsubTest : IDisposable
             Command = "Pubsub"
         };
 
+        /// <summary>
+        /// Handle a special race condition that can occur when deleting
+        /// something:
+        /// 1. Delete request times out.
+        /// 2. Delete operation continues on server and succeeds.
+        /// 3. Later requests to delete the same entity see NotFound error.
+        /// </summary>
+        /// <param name="delete">The delete operation to run.</param>
+        /// <returns>An action to run inside Eventually().</returns>
+        Action HandleDeleteRace(Action delete)
+        {
+            bool sawTimeout = false;
+            return () =>
+            {
+                if (!sawTimeout)
+                {
+                    try
+                    {
+                        delete();
+                    }
+                    catch (Grpc.Core.RpcException e)
+                    when (e.Status.StatusCode == StatusCode.DeadlineExceeded)
+                    {
+                        sawTimeout = true;
+                        throw;
+                    }
+                }
+                else
+                {
+                    try
+                    {
+                        delete();
+                    }
+                    catch (Grpc.Core.RpcException e)
+                    when (e.Status.StatusCode == StatusCode.NotFound)
+                    {
+                        // Earlier timeout request that deleted the thing
+                        // actually succeeded on the server.
+                    }
+                }
+            };
+        }
+
         void IDisposable.Dispose()
         {
             foreach (string topicId in _tempTopicIds)
             {
-                Eventually(() => Run("deleteTopic", _projectId, topicId));
+                Eventually(HandleDeleteRace(() =>

```

Comment:
We should not throw exceptions in Dispose method - even if it's a sample code.
Dispose is executed in a finally block which is usually executed in the context of resources cleanup.
https://msdn.microsoft.com/en-us/library/bb386039.aspx
Also, Dispose should implement a logic to prevent freeing resources twice (if it's called by mistake twice) and still, even if called twice, it shouldn't throw exception.
https://msdn.microsoft.com/en-us/library/system.idisposable.dispose(v=vs.110).aspx

In your case HandleDeleteRace can throw exception.

Reply:
xUnit reuses/abuses the `IDisposable` interface for the purposes of test clean up.  It's described here: https://xunit.github.io/docs/shared-context.html

I wish xUnit had defined a separate interface instead of reusing `IDisposable` so it would be less confusing.  In short, the usual rules for `Dispose()` are overkill and do not apply to xUnit tests.  

In summary:
1. The xUnit framework is the only thing that will ever call this `Dispose()` method.
2. The xUnit framework is will call the method exactly once.
3. The xUnit framework is will recover if an exception is thrown.

Reply:
Yeah, I just saw that and I really don't like what they did - they say to initialize everything per test in constructor, which is also antipattern and they call it idiomatic solution... If exception is thrown, the test framework catches that but the callstack and result reporting is misleading. The problem is that your test could succeed, test cleanup failed and it will show that the test failed with callstack from Dispose(). There should be clear distinction between test failure vs. test setup/cleanup failure.  The only upside from abusing Dispose and doing initialization in c-tor is that you can mark fields as readonly.

Anyway, I don't want to block you as I understand that the problem stems from xUnit.
I'll try to think about a way to make xUnit test initialization more tester friendly without abusing language features.